### ISSUE TITLE
Parallel Process

### ISSUE THREAD

**Issue by wilkerlucio** [wilkerlucio/pathom3#20]
Pathom 3 ability to run resolvers in parallel.

Compared to Pathom 2, the Pathom 3 parallel implementation is likely to be simpler. The new planner results already have all the information the runner needs to know what it can parallelize, this will reduce the large amount of complexity that was required in the parallel runner from Pathom 2.

The challenge now becomes how to do good resource management around it. Given Pathom 3 uses Promesa, the initial feeling is that we could just let it run. But in my experience, this can easily turn into high resource usage on the server, so I like to add some control mechanisms in Pathom.

Some tunning dials I like to offer in Pathom to control parallelism:

- [ ] Maximum number of collection items to run in parallel in a given process
- [ ] Maximum number of parallel resolvers to run sideways
- [ ] Maximum number of threads used per request

And some current things I know are broken in parallel and need fixing (not a bug for current usage, since there is not parallel process):

- [ ] Ensure parallel merging doesn't lose data in concurrent operation s

If anybody has experience with high concurrency on the JVM (especially using `CompletableFuture`), I would love to know about ideas to make an efficient engine for a parallel process.

**Comment by cyppan** [comment]
I'm currently trying the asynchronous eql processing (p.a.eql/process), what do you mean by "Ensure parallel merging doesn't lose data in concurrent operations"? What is the bug exactly?

**Comment by wilkerlucio** [comment]
Hello @cyppan, this ticket is to hold information for a future feature of the parallel process. Parallel merging is not an issue now because there is no parallel process, this is just a note I made for myself because of things I see when I tried to play with this before. But this is on hold, and I should get back to that once the dynamic parts are ready and documented.

**Comment by cyppan** [comment]
Ok 👍

**Comment by mauricioszabo** [comment]
I think Promesa already have a way to control the connection pool, generate new pools, etc... did you try to play with the knobs on Promesa? (or, the problem is more complicated and I'm waaaay oversimplificating things, which is quite possible too? :laughing:)
